### PR TITLE
Implemented relative resolution (rfc3986)

### PR DESCRIPTION
**Pull request by umbreak**
[rfc3986](https://tools.ietf.org/html/rfc3986#section-5.2)

### CODE REVIEW

File: modules/core/src/test/scala/ch/epfl/bluebrain/nexus/rdf/PathSpec.scala [hunk 1]
```diff
@@ -92,5 +92,30 @@ class PathSpec extends WordSpecLike with Matchers with Inspectors with EitherVal
...
+    }
+
+    "does not start with slash" in {
+      val cases = List(Empty, Segment("d", Slash(Segment("c", Slash(Slash(Segment("b", Slash(Segment("a", Empty)))))))))
+      forAll(cases) {
+        case (p) => p.startWithSlash shouldEqual false
+      }
+    }
+
+    "reverse" in {
```
Comment (bogdanromanx) [inline modules/core/src/test/scala/ch/epfl/bluebrain/nexus/rdf/PathSpec.scala]: add tests:
*   when path is empty
*   when path ends with a slash
*   when path is just a slash

File: modules/core/src/test/scala/ch/epfl/bluebrain/nexus/rdf/RelativeIriSpec.scala [hunk 2]
```diff
@@ -79,5 +79,35 @@ class RelativeIriSpec extends WordSpecLike with Matchers with Inspectors with Ei
...
+        "g;x"     -> "http://a/b/c/g;x",
+        "g;x?y#s" -> "http://a/b/c/g;x?y#s",
+        "."       -> "http://a/b/c/",
+        "./"      -> "http://a/b/c/",
+        ".."      -> "http://a/b/",
+        "../"     -> "http://a/b/",
+        "../g"    -> "http://a/b/g",
+        "../.."   -> "http://a/",
+        "../../"  -> "http://a/",
+        "../../g" -> "http://a/g"
```
Comment (bogdanromanx) [inline modules/core/src/test/scala/ch/epfl/bluebrain/nexus/rdf/RelativeIriSpec.scala]: add additional test cases:
*   when base ends with slash
*   when base contains a fragment
*   when the number of `..` segments is higher than the number of segments in the base
*   combine `.` and `..` to see them collapsed at the same time

File: modules/core/src/test/scala/ch/epfl/bluebrain/nexus/rdf/PathSpec.scala [hunk 3]
```diff
@@ -92,5 +92,30 @@ class PathSpec extends WordSpecLike with Matchers with Inspectors with EitherVal
...
+      Path("/a/b/c/d").right.value.reverse shouldEqual Slash(
+        Segment("a", Slash(Segment("b", Slash(Segment("c", Slash(Segment("d", Empty))))))))
+      Path("/a/b/c/d").right.value.reverse.reverse shouldEqual Path("/a/b/c/d").right.value
+    }
+
+    "join two paths" in {
```
Comment (bogdanromanx) [inline modules/core/src/test/scala/ch/epfl/bluebrain/nexus/rdf/PathSpec.scala]: add additional tests: use a mixture of empty, slash only, ending with slash paths.

File: modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Iri.scala [hunk 4]
```diff
@@ -747,6 +826,38 @@ object Iri {
...
+      * @param other the path to be prepended to the current path
+      * @return current / suffix
+      */
+    def prepend(other: Path): Path
+
+    /**
+      * @param segment the segment to be appended to a path
+      * @return a new Path with the ending segment
+      */
+    def ::(segment: String): Path
```
[the remark this answers is not in the document]
Comment (bogdanromanx) [reply modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Iri.scala]: and inject the `/`

File: modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Iri.scala [hunk 5]
```diff
@@ -160,6 +161,79 @@ object Iri {
...
+      }
+
+    private def resolveUrl(base: Url): AbsoluteIri =
+      authority match {
+        case Some(auth) =>
+          base.copy(authority = Some(auth), query = query, path = path, fragment = fragment)
+        case None =>
+          val (p, q) = path match {
+            case Empty =>
+              base.path -> (if (query.isDefined) query else base.query)
```
[the remark this answers is not in the document]
Comment (umbreak) [reply modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Iri.scala]: Yes. From the RFC
```
            if (R.path == "") then
               T.path = Base.path;
               if defined(R.query) then
                  T.query = R.query;
               else
                  T.query = Base.query;
               endif;
            else 
               ...
               T.query = R.query;
```

File: modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Iri.scala [hunk 6]
```diff
@@ -160,6 +161,79 @@ object Iri {
...
+
+    private def merge(base: Url): Path =
+      if (base.authority.isDefined && base.path.isEmpty) Slash(path)
+      else removeDotSegments(path :: deleteLast(base.path))
+
+    private def merge(base: Urn): Path =
+      removeDotSegments(path :: deleteLast(base.nss))
+
+    private def deleteLast(path: Path, withSlash: Boolean = false): Path =
+      path match {
```
[the remark this answers is not in the document]
Comment (umbreak) [reply modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Iri.scala]: The RFC talks about `remove the last segment and the and its preceding "/" (if any) from the output buffer`. It does not say to remove a "/" straight away.

File: modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Iri.scala [hunk 7]
```diff
@@ -825,6 +944,10 @@ object Iri {
       def asSegment: Option[Segment] = Some(this)
       def asString: String           = rest.asString + segment
       def pctEncoded: String         = rest.pctEncoded + pctEncode(segment)
+      def startWithSlash: Boolean    = rest.startWithSlash
+      def prepend(other: Path): Path = segment :: (rest prepend other)
+      def ::(s: String): Path        = if (segment.isEmpty) this else Segment(segment + s, rest)
```
[the remark this answers is not in the document]
Comment (umbreak) [reply modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Iri.scala]: Yes. It is how akka is doing it and anything else is an assumption that the client did something wrong (which is what we do in the `Path` implementation in service).

File: modules/core/src/test/scala/ch/epfl/bluebrain/nexus/rdf/RelativeIriSpec.scala (hunk 2 above)
Comment (umbreak) [reply modules/core/src/test/scala/ch/epfl/bluebrain/nexus/rdf/RelativeIriSpec.scala]: The tests are taken form the RFC. But yes, I can add those others

File: modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Iri.scala [hunk 8]
```diff
@@ -945,8 +960,8 @@ object Iri {
       def asString: String           = rest.asString + segment
       def pctEncoded: String         = rest.pctEncoded + pctEncode(segment)
       def startWithSlash: Boolean    = rest.startWithSlash
-      def prepend(other: Path): Path = segment :: (rest prepend other)
-      def ::(s: String): Path        = if (segment.isEmpty) this else Segment(segment + s, rest)
+      def prepend(other: Path): Path = (rest prepend other) / segment
+      def /(s: String): Path         = if (segment.isEmpty) this else Segment(segment + s, rest)
```
Comment (bogdanromanx) [inline modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Iri.scala]: are we sure we want to do this? it implies this behaviour:
` "/a/b/c" / "d" => "/a/b/cd"`
and I'd expect:
` "/a/b/c" / "d" => "/a/b/c/d"`
